shell/platform/windows/flutter_window.cc

```diff
@@ -0,0 +1,286 @@
+
+  internal_plugin_registrar_ =
+      std::make_unique<flutter::PluginRegistrar>(plugin_registrar_.get());
+
+  // Set up the keyboard handlers.
+  auto internal_plugin_messenger = internal_plugin_registrar_->messenger();
+  keyboard_hook_handlers_.push_back(
+      std::make_unique<flutter::KeyEventHandler>(internal_plugin_messenger));
+  keyboard_hook_handlers_.push_back(
+      std::make_unique<flutter::TextInputPlugin>(internal_plugin_messenger));
+  platform_handler_ = std::make_unique<flutter::PlatformHandler>(
+      internal_plugin_messenger, this);
+
+  auto state = std::make_unique<FlutterDesktopWindowControllerState>();
+  state->engine = engine_;
+  state->window = nullptr;  // null here as the caller owns the uniqueptr we'll
+                            // ultimately want to return
+  state->window_wrapper =
+      std::make_unique<FlutterDesktopWindow>();  // a different window wrapper
+                                                 // for same window pointer
+                                                 // hence make_unique again
+  state->window_wrapper->window = this;
+  return state.release();
+}
+
+FlutterDesktopPluginRegistrarRef FlutterWindow::GetRegistrar() {
+  return plugin_registrar_.get();
+}
+
+void FlutterWindow::HandleMessage(const FlutterDesktopMessage& message,
+                                  std::function<void(void)> input_block_cb,
+                                  std::function<void(void)> input_unblock_cb) {
+  message_dispatcher_->HandleMessage(message, input_block_cb, input_unblock_cb);
+}
+
+// When DesktopWindow notifies that a WM_Size message has come in
+// let FlutterEngine know about the new size.
+void FlutterWindow::OnResize(unsigned int width, unsigned int height) {
+  current_width_ = width;
+  current_height_ = height;
+  SendWindowMetrics();
+}
+// When DesktopWindow notifies that a WM_Size message has come in
+// let FlutterEngine know about the new size.
+void FlutterWindow::OnResize() {
+  SendWindowMetrics();
+}
+
+// When DesktopWindow notifies that the DPI has changed
+// handle the change.
+void FlutterWindow::OnDpiScale(unsigned int dpi) {
+  current_dpi_ = dpi;
+
+  // No need to SendWindowMetrics here as OnResize will be called right after
+  // DPI change
+}
+
+void FlutterWindow::OnPointerMove(double x, double y) {
+  SendPointerMove(x, y);
+}
+
+void FlutterWindow::OnPointerDown(double x, double y) {
+  SendPointerDown(x, y);
+}
+
+void FlutterWindow::OnPointerUp(double x, double y) {
+  SendPointerUp(x, y);
+}
+
+void FlutterWindow::OnChar(unsigned int code_point) {
+  SendChar(code_point);
+}
+
+void FlutterWindow::OnKey(int key, int scancode, int action, int mods) {
+  SendKey(key, scancode, action, 0);
+}
+
+void FlutterWindow::OnScroll(double delta_x, double delta_y) {
+  SendScroll(delta_x, delta_y);
+}
+
+void FlutterWindow::FlutterMessageLoop() {
+  MSG message;
+
+  while (GetMessage(&message, nullptr, 0, 0)) {
```

Comment:
This is blocking, right? If so, this message loop is going to have the same problem as described in the mailing list message I linked to in earlier emails, where Flutter plugin messages aren't going to be received until some native event like a mouse move happens. I would rather busy-wait with a non-blocking version for now; high CPU is a better regression than incorrect and confusing behavior.

Reply:
(Also, this needs a TODO about using a full message loop implementation.)

Reply:
Disregard the comment about switching to a busy wait. It's not a big deal either way for landing the initial version, and it'll be substantially rewritten later anyway so iterating on it now would be wasted effort.

I filed https://github.com/flutter/flutter/issues/36420 so you can just add a TODO referencing that and leave the code as-is.